{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "28341458",
   "metadata": {},
   "source": [
    "# Molecular Oncology Almanac workshop\n",
    "This notebook is being used as a demonstration for the [Cancer Genomics Consortium 2023](https://cancergenomics.org/meetings/cgc_annual_meeting_2023.php) Bioinformatics Workshop on Exploring the Clinical Interpretation Resource Landscape. It can be found on GitHub at [github.com/vanallenlab/2023-cgc-moalmanac](https://github.com/vanallenlab/2023-cgc-moalmanac) and you can learn more at [moalmanac.org](moalmanac.org).\n",
    "\n",
    "**For this notebook, please move it to your moalmanac/moalmanac execution directory to import modules!**\n",
    "https://github.com/vanallenlab/moalmanac"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ea7df90",
   "metadata": {},
   "source": [
    "<img src=\"img/moalmanac-browser.png\" alt=\"Molecular Oncology Almanac browser\" width=\"750\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e0aa0db",
   "metadata": {},
   "source": [
    "<a id='toc'></a>\n",
    "## Table of contents\n",
    "- <a href=\"#algorithm\">Example algorithm execution</a>\n",
    "    - <a href=\"#configure\">Configuration</a>\n",
    "    - <a href=\"#inputs\">Specify inputs</a>\n",
    "    - <a href=\"#run\">Run algorithm</a>\n",
    "    - <a href=\"#outputs\">Review outputs</a>\n",
    "- <a href=\"#conclusion\">Conclusion</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bcc12004",
   "metadata": {},
   "source": [
    "<a id='algorithm'></a>\n",
    "## Algorithm execution\n",
    "The codebase for the Molecular Oncology Almanac algorithm can be found on [GitHub](https://github.com/vanallenlab/moalmanac) or [Docker Hub](https://hub.docker.com/r/vanallenlab/moalmanac/). \n",
    "\n",
    "<img src=\"img/moalmanac-algorithm-schematic.png\" alt=\"Molecular Oncology Almanac browser\" width=\"750\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dc1e082",
   "metadata": {},
   "source": [
    "# Please move this notebook to your moalmanac/moalmanac execution directory!\n",
    "Please download and set up MOAlmanac (https://github.com/vanallenlab/moalmanac) for use on your local machine to run this notebook. **This notebook should be executed from your moalmanac/moalmanac directory**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "56a031fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "python_path = '/'.join(sys.executable.split('/')[:-1])\n",
    "pip_path = f\"{python_path}/pip\"\n",
    "\n",
    "\n",
    "virtual_env_name = \"2023-cgc-workshop-moalmanac\"\n",
    "moalmanac_directory = '/Users/brendan/Github/moalmanac/moalmanac'\n",
    "\n",
    "! $pip_path install -q -r $moalmanac_directory/requirements.txt\n",
    "\n",
    "! cd /Users/brendan/Github/moalmanac/moalmanac\n",
    "\n",
    "import moalmanac"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e28d1024",
   "metadata": {},
   "source": [
    "<a id='configure'></a>\n",
    "### Configure implementation\n",
    "MOAlmanac contains two configuration files to customize of the algorithm's use. \n",
    "- `config.ini` allows users to change settings of the algorithm and toggle features\n",
    "- `colnames.ini` allows users to change strings expected to be found within input data, internally to the algorithm, and written as outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "c80f9fda",
   "metadata": {},
   "outputs": [],
   "source": [
    "from config import CONFIG\n",
    "from config import COLNAMES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "610d2d7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "calculate_model_similarity\n",
      "calculate_mutational_signatures\n",
      "calculate_preclinical_efficacy\n",
      "generate_actionability_report\n",
      "include_model_similarity_in_actionability_report\n",
      "include_preclinical_efficacy_in_actionability_report\n",
      "plot_mutational_signatures\n",
      "plot_preclinical_efficacy\n"
     ]
    }
   ],
   "source": [
    "for key in CONFIG['function_toggle']:\n",
    "    print(key)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0630fcd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['input_data', 'tcga_maf_input', 'gdc_maf_input', 'indel_input', 'germline_input', 'validation_input', 'seg_input', 'called_cn_input', 'fusion_input', 'simple_input', 'maps', 'patient', 'aneuploidy', 'features', 'burden', 'microsatellite', 'signatures', 'oncotree', 'validation_sequencing', 'overlap_somatic_germline', 'datasources', 'integrative', 'bin_names', 'report', 'outputs', 'preclinical', 'matchmaking'])"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COLNAMES.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "faece539",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'gene': 'hugo_symbol',\n",
       " 'chr': 'chromosome',\n",
       " 'start': 'start_position',\n",
       " 'end': 'end_position',\n",
       " 'ref': 'reference_allele',\n",
       " 'allele1': 'tumor_seq_allele1',\n",
       " 'allele2': 'tumor_seq_allele2',\n",
       " 'alt_type': 'variant_classification',\n",
       " 'alt': 'protein_change',\n",
       " 'tumor': 'tumor_sample_barcode',\n",
       " 'normal': 'matched_norm_sample_barcode',\n",
       " 'ref_count': 't_ref_count',\n",
       " 'alt_count': 't_alt_count'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "COLNAMES['germline_input']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "09625cc2",
   "metadata": {},
   "source": [
    "<a id='inputs'></a>\n",
    "### Specify inputs\n",
    "The Molecular Oncology Almanac will annotate and evaluate any combination of inputs provided.Complete documentation of inputs accepted is [available within the algorithm's documentation](https://github.com/vanallenlab/moalmanac/blob/main/docs/description-of-inputs.md). \n",
    "\n",
    "In short for sample metadata,\n",
    "- a profile id with `--patient_id` is the only required argument.\n",
    "- a tumor type can be provided with `--tumor_type`, which will be mapped to [OncoTree](https://oncotree.mskcc.org/#/home) codes.  \n",
    "- additional metadata can be passed with `--stage` and `--description`, both free text fields.\n",
    "\n",
    "and for genomic alterations,\n",
    "- somatic variants and germline variants can be provided with `--snv_handle`, `--indel_handle`, and `--germline_handle` as MAF files following either NCI or TCGA specification.\n",
    "- tumor mutational burden will be calculated if the bases covered to call somatic variants is provided as an integer with the argument `--bases_covered_handle\n",
    "- called gene-level copy number alterations can be provided with `--called_cn_handle` and fusions can be provided with `--fusion_handle`. \n",
    "- purity and ploidy can be provided as float values with `--purity` and `--ploidy` as well. \n",
    "- somatic variants called from validation sequencing can be provided with `--validation_handle`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ec55d6b",
   "metadata": {},
   "source": [
    "Alternatively, a simplified input file that contains minimal variant details may be provided to the algorithm. Such as,\n",
    "\n",
    "| feature_type | feature | alteration_type | alteration |\n",
    "|---|---|---|---|\n",
    "| Somatic Variant | BRAF | Missense | p.V600E |\n",
    "| Copy Number | CDK4 | Amplification |  |\n",
    "| Rearrangement | COL1A1 | Fusion | COL1A1--CITED4 |\n",
    "| Germline Variant | BRCA2 | Frameshift | p.S1982fs |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0da9d5ac",
   "metadata": {},
   "source": [
    "Additional settings can be set by modifying the config.ini file and column names of provided input files may be modified by editing the colnames.ini file.\n",
    "\n",
    "Required arguments:\n",
    "```\n",
    "    --patient_id            <string>    patient identifier\n",
    "```\n",
    "\n",
    "Optional arguments:\n",
    "```\n",
    "    --tumor_type            <string>    tumor ontology, default=Unknown\n",
    "    --stage                 <string>    tumor stage, default=Unknown\n",
    "    --snv_handle            <string>    handle for MAF file of somatic single nucleotide variants\n",
    "    --indel_handle          <string>    handle for MAF file of somatic insertions and deletions\n",
    "    --bases_covered_handle  <string>    handle for text file which contains the number of calcable somatic bases\n",
    "    --called_cn_handle      <string>    handle for text file which contained genes and copy number calls, will be used over `--cnv_handle`\n",
    "    --cnv_handle            <string>    handle for annotated seg file for somatic copy number\n",
    "    --fusion_handle         <string>    handle for STAR fusion output, .final.abridged\n",
    "    --germline_handle       <string>    handle for MAF file of germline single nucleotide variants and insertions and deletions\n",
    "    --validation_handle     <string>    handle for MAF file of somatic single nucleotide variant called from validation sequencing\n",
    "    --ms_status             <string>    microsatellite status as deemed by MSI sensor, MSI or MSS, default=Unknown\n",
    "    --purity                <float>     tumor purity\n",
    "    --ploidy                <float>     tumor ploidy\n",
    "    --wgd                   <boolean>   specify the occurence of whole genome duplication\n",
    "    --description           <string>    description of patient\n",
    "    --output-directory      <string>    specify location of produced outputs\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e79e960",
   "metadata": {},
   "source": [
    "Here, we will utilize example data found within in this repository for use with moalmanac. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c1ec4253",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_dictionary = {\n",
    "    'patient_id': 'example',\n",
    "    'reported_tumor_type': 'SKCM',\n",
    "    'stage': 'Metastatic',\n",
    "    'description': 'Example data for CGC workshop!',\n",
    "    'purity': 0.85,\n",
    "    'ploidy': 4.02,\n",
    "    'WGD': True,\n",
    "    'microsatellite_status': 'msih'\n",
    "}\n",
    "\n",
    "root = \"/Users/brendan/Github/scratch/2023-cgc-moalmanac/\"\n",
    "input_dictionary = {\n",
    "    'snv_handle': '/Users/brendan/Github/scratch/2023-cgc-moalmanac/example-data/example_patient.capture.somatic.snvs.maf',\n",
    "    'indel_handle': '/Users/brendan/Github/scratch/2023-cgc-moalmanac/example-data/example_patient.capture.somatic.indels.maf',\n",
    "    'bases_covered_handle': '/Users/brendan/Github/scratch/2023-cgc-moalmanac/example-data/example_patient.capture.somatic.coverage.txt',\n",
    "    'called_cn_handle': '/Users/brendan/Github/scratch/2023-cgc-moalmanac/example-data/example_patient.capture.somatic.called.cna.txt',\n",
    "    'cnv_handle': '',\n",
    "    'fusion_handle': '/Users/brendan/Github/scratch/2023-cgc-moalmanac/example-data/example_patient.rna.star.fusions.txt',\n",
    "    'germline_handle': '/Users/brendan/Github/scratch/2023-cgc-moalmanac/example-data/example_patient.capture.germline.maf',\n",
    "    'validation_handle': '/Users/brendan/Github/scratch/2023-cgc-moalmanac/example-data/example_patient.rna.somatic.snvs.maf',\n",
    "    'disable_matchmaking': True\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74f14c9d",
   "metadata": {},
   "source": [
    "<a id='run'></a>\n",
    "### Run algorithm\n",
    "The Molecular Oncology Almanac algorithm can be executed by running the Python code from Bash, from within a Python environment by instantiating the main function, or by utilizing or modifying the [example run script](https://github.com/vanallenlab/moalmanac/blob/main/moalmanac/run_example.py). Please execute [moalmanac/simplified_input.py](https://github.com/vanallenlab/moalmanac/tree/main/moalmanac#simplified-input) if performing interpretation on a simpified input. Otherwise, an example execution may appear as,\n",
    "\n",
    "```\n",
    "python moalmanac.py \\\n",
    "    --patient_id \"example\" \\\n",
    "    --tumor_type \"SKCM\" \\\n",
    "    --stage \"Metastatic\" \\\n",
    "    --description \"Example profile for interpretation with the Molecular Oncology Almanac\" \\\n",
    "    --snv_handle \"example-data/example_patient.capture.somatic.snvs.maf\" \\\n",
    "    --indel_handle \"example-data/example_patient.capture.somatic.indels.maf\" \\\n",
    "    --bases_covered_handle \"example-data/example_patient.capture.somatic.coverage.txt\" \\\n",
    "    --called_cn_handle \"example-data/example_patient.capture.somatic.called.cna.txt\" \\\n",
    "    --fusion_handle \"example-data/example_patient.capture.somatic.seg.annotated\" \\\n",
    "    --germline_handle \"example-data/example_patient.rna.star.fusions.txt\" \\\n",
    "    --validation_handle \"example-data/example_patient.rna.somatic.snvs.maf\" \\\n",
    "    --purity 0.85 \\\n",
    "    --ploidy 4.02 \\\n",
    "    --ms_status \"msih\" \\\n",
    "    --wgd\n",
    "```\n",
    "\n",
    "Interpretation on clinically or biologically relevant genomic alterations cause the computational run time to scale linearly. Using the example data in this repository, performing just the somatic and germline interpretation takes ~90 seconds to complete (19 clinically or biologically relevant variants, along with the rest of the provided variants). This time will increase if you enable: report generation, preclinical efficacy, similarity comparisons, and mutational signature calculations. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9af46be5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Molecular Oncology Almanac runtime: 92.05 seconds\n"
     ]
    }
   ],
   "source": [
    "output_directory = \"workshop-outputs/\"\n",
    "!mkdir -p $output_directory\n",
    "\n",
    "import time\n",
    "start_time = time.time()\n",
    "\n",
    "moalmanac.main(metadata_dictionary, input_dictionary, output_directory)\n",
    "\n",
    "end_time = time.time()\n",
    "time_statement = \"Molecular Oncology Almanac runtime: %s seconds\" % round((end_time - start_time), 4)\n",
    "print(time_statement)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bfaa6910",
   "metadata": {},
   "source": [
    "<a id='outputs'></a>\n",
    "### Review outputs\n",
    "The Molecular Oncology Almanac has the ability to produce several outputs, dependant on which inputs are provided. In all cases, a [fully-contained interactive report](example-outputs/example.report.html) will be generated. Complete documentation of outputs produced is [available within the algorithm's documentation](https://github.com/vanallenlab/moalmanac/blob/main/docs/description-of-outputs.md)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "bceb4c22",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f1524e3",
   "metadata": {},
   "source": [
    "Genomic alterations that are associated with clinical or biological significance are highlighted in the `.actionable.txt` output. These are the same findings that will be reported within the [interactive report](example-outputs/example.report.html).\n",
    "\n",
    "This output contains one row for each genomic alteration highlighted along with associated relevance for therapeutic sensitivity, resistance, or disease prognosis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "50cf52ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score_bin</th>\n",
       "      <th>sensitive_predictive_implication</th>\n",
       "      <th>resistance_predictive_implication</th>\n",
       "      <th>prognostic_predictive_implication</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>alteration_type</th>\n",
       "      <th>alteration</th>\n",
       "      <th>tumor_f</th>\n",
       "      <th>total_coverage</th>\n",
       "      <th>...</th>\n",
       "      <th>prognostic_url</th>\n",
       "      <th>number_germline_mutations_in_gene</th>\n",
       "      <th>validation_total_coverage</th>\n",
       "      <th>validation_tumor_f</th>\n",
       "      <th>validation_detection_power</th>\n",
       "      <th>feature_display</th>\n",
       "      <th>preclinical_efficacy_observed</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>tumor_sample_barcode</th>\n",
       "      <th>normal_sample_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>FDA-Approved</td>\n",
       "      <td>Guideline</td>\n",
       "      <td>Guideline</td>\n",
       "      <td>Somatic Variant</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>Missense</td>\n",
       "      <td>p.V600E</td>\n",
       "      <td>0.6316</td>\n",
       "      <td>152.0</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.nccn.org/professionals/physician_g...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.161</td>\n",
       "      <td>BRAF p.V600E (Missense)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>example</td>\n",
       "      <td>example_tumor_profile</td>\n",
       "      <td>example_normal_profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Investigate Actionability</td>\n",
       "      <td>FDA-Approved</td>\n",
       "      <td>Guideline</td>\n",
       "      <td>Guideline</td>\n",
       "      <td>Microsatellite Stability</td>\n",
       "      <td>MSI-High</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>https://www.nccn.org/professionals/physician_g...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MSI-High</td>\n",
       "      <td>NaN</td>\n",
       "      <td>example</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Investigate Actionability</td>\n",
       "      <td>Guideline</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Rearrangement</td>\n",
       "      <td>COL1A1</td>\n",
       "      <td>Fusion</td>\n",
       "      <td>COL1A1--CITED4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>COL1A1--CITED4 Fusion</td>\n",
       "      <td>0.0</td>\n",
       "      <td>example</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Investigate Actionability</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Inferential</td>\n",
       "      <td>Aneuploidy</td>\n",
       "      <td>Whole genome doubling</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>https://doi.org/10.1038/s41588-018-0165-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>example</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Investigate Actionability</td>\n",
       "      <td>FDA-Approved</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Germline Variant</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>Frameshift</td>\n",
       "      <td>p.S1982fs</td>\n",
       "      <td>0.5000</td>\n",
       "      <td>100.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>BRCA2 p.S1982fs (Frameshift)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>example</td>\n",
       "      <td>example_tumor_profile</td>\n",
       "      <td>example_normal_profile</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   score_bin sensitive_predictive_implication   \n",
       "0      Putatively Actionable                     FDA-Approved  \\\n",
       "1  Investigate Actionability                     FDA-Approved   \n",
       "2  Investigate Actionability                        Guideline   \n",
       "3  Investigate Actionability                              NaN   \n",
       "4  Investigate Actionability                     FDA-Approved   \n",
       "\n",
       "  resistance_predictive_implication prognostic_predictive_implication   \n",
       "0                         Guideline                         Guideline  \\\n",
       "1                         Guideline                         Guideline   \n",
       "2                               NaN                               NaN   \n",
       "3                               NaN                       Inferential   \n",
       "4                               NaN                               NaN   \n",
       "\n",
       "               feature_type                feature alteration_type   \n",
       "0           Somatic Variant                   BRAF        Missense  \\\n",
       "1  Microsatellite Stability               MSI-High             NaN   \n",
       "2             Rearrangement                 COL1A1          Fusion   \n",
       "3                Aneuploidy  Whole genome doubling             NaN   \n",
       "4          Germline Variant                  BRCA2      Frameshift   \n",
       "\n",
       "       alteration  tumor_f  total_coverage  ...   \n",
       "0         p.V600E   0.6316           152.0  ...  \\\n",
       "1             NaN      NaN             NaN  ...   \n",
       "2  COL1A1--CITED4      NaN             NaN  ...   \n",
       "3             NaN      NaN             NaN  ...   \n",
       "4       p.S1982fs   0.5000           100.0  ...   \n",
       "\n",
       "                                      prognostic_url   \n",
       "0  https://www.nccn.org/professionals/physician_g...  \\\n",
       "1  https://www.nccn.org/professionals/physician_g...   \n",
       "2                                                NaN   \n",
       "3          https://doi.org/10.1038/s41588-018-0165-1   \n",
       "4                                                NaN   \n",
       "\n",
       "   number_germline_mutations_in_gene validation_total_coverage   \n",
       "0                                1.0                       4.0  \\\n",
       "1                                NaN                       NaN   \n",
       "2                                NaN                       NaN   \n",
       "3                                NaN                       NaN   \n",
       "4                                NaN                       NaN   \n",
       "\n",
       "  validation_tumor_f validation_detection_power               feature_display   \n",
       "0                0.0                      0.161       BRAF p.V600E (Missense)  \\\n",
       "1                NaN                        NaN                      MSI-High   \n",
       "2                NaN                        NaN         COL1A1--CITED4 Fusion   \n",
       "3                NaN                        NaN                           NaN   \n",
       "4                NaN                        NaN  BRCA2 p.S1982fs (Frameshift)   \n",
       "\n",
       "  preclinical_efficacy_observed patient_id   tumor_sample_barcode   \n",
       "0                           1.0    example  example_tumor_profile  \\\n",
       "1                           NaN    example                    NaN   \n",
       "2                           0.0    example                    NaN   \n",
       "3                           NaN    example                    NaN   \n",
       "4                           NaN    example  example_tumor_profile   \n",
       "\n",
       "    normal_sample_barcode  \n",
       "0  example_normal_profile  \n",
       "1                     NaN  \n",
       "2                     NaN  \n",
       "3                     NaN  \n",
       "4  example_normal_profile  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = 'example-outputs/example.actionable.txt'\n",
    "df = pd.read_csv(file, sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84931d19",
   "metadata": {},
   "source": [
    "Somatic nucleotide variants (`--snv_handle` and (`--indel_handle`), copy number alterations (`--called_cn_handle`), and fusions (`--`) are summarized in the the `.somatic.scored.txt` output. This output reports a broad clinical interpretation and biological interpretation by presence or absence in several datasources. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f1f5b1cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score_bin</th>\n",
       "      <th>sensitive_score_bin</th>\n",
       "      <th>resistance_score_bin</th>\n",
       "      <th>prognostic_score_bin</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>alteration_type</th>\n",
       "      <th>alteration</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start_position</th>\n",
       "      <th>...</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>tumor_sample_barcode</th>\n",
       "      <th>normal_sample_barcode</th>\n",
       "      <th>almanac_bin</th>\n",
       "      <th>cancerhotspots_bin</th>\n",
       "      <th>cancerhotspots3D_bin</th>\n",
       "      <th>cgc_bin</th>\n",
       "      <th>gsea_pathways_bin</th>\n",
       "      <th>gsea_modules_bin</th>\n",
       "      <th>cosmic_bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>Somatic Variant</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>Missense</td>\n",
       "      <td>p.V600E</td>\n",
       "      <td>7</td>\n",
       "      <td>140453136.0</td>\n",
       "      <td>...</td>\n",
       "      <td>example</td>\n",
       "      <td>example_tumor_profile</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>CDK4</td>\n",
       "      <td>Amplification</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>example</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>example</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>BRAF</td>\n",
       "      <td>Amplification</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>example</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>Putatively Actionable</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>CDKN2A</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>example</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               score_bin    sensitive_score_bin   resistance_score_bin   \n",
       "0  Putatively Actionable  Putatively Actionable  Putatively Actionable  \\\n",
       "1  Putatively Actionable  Putatively Actionable                    NaN   \n",
       "2  Putatively Actionable                    NaN                    NaN   \n",
       "3  Putatively Actionable                    NaN  Putatively Actionable   \n",
       "4  Putatively Actionable  Putatively Actionable                    NaN   \n",
       "\n",
       "    prognostic_score_bin     feature_type feature alteration_type alteration   \n",
       "0  Putatively Actionable  Somatic Variant    BRAF        Missense    p.V600E  \\\n",
       "1                    NaN      Copy Number    CDK4   Amplification        NaN   \n",
       "2  Putatively Actionable      Copy Number    TP53        Deletion        NaN   \n",
       "3                    NaN      Copy Number    BRAF   Amplification        NaN   \n",
       "4                    NaN      Copy Number  CDKN2A        Deletion        NaN   \n",
       "\n",
       "  chromosome  start_position  ...  patient_id   tumor_sample_barcode   \n",
       "0          7     140453136.0  ...     example  example_tumor_profile  \\\n",
       "1        NaN             NaN  ...     example                    NaN   \n",
       "2        NaN             NaN  ...     example                    NaN   \n",
       "3        NaN             NaN  ...     example                    NaN   \n",
       "4        NaN             NaN  ...     example                    NaN   \n",
       "\n",
       "    normal_sample_barcode almanac_bin  cancerhotspots_bin   \n",
       "0  example_normal_profile           4                   2  \\\n",
       "1                     NaN           4                   1   \n",
       "2                     NaN           4                   1   \n",
       "3                     NaN           4                   1   \n",
       "4                     NaN           4                   1   \n",
       "\n",
       "   cancerhotspots3D_bin  cgc_bin  gsea_pathways_bin  gsea_modules_bin   \n",
       "0                     2        1                  1                 1  \\\n",
       "1                     2        1                  1                 1   \n",
       "2                     1        1                  1                 1   \n",
       "3                     1        1                  1                 1   \n",
       "4                     0        1                  1                 1   \n",
       "\n",
       "   cosmic_bin  \n",
       "0           2  \n",
       "1           1  \n",
       "2           1  \n",
       "3           1  \n",
       "4           1  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = 'example-outputs/example.somatic.scored.txt'\n",
    "df = pd.read_csv(file, sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb2d61ff",
   "metadata": {},
   "source": [
    "If germline variants are provided using the `--germline_handle`, outputs will be produced to highlight variants in genes: \n",
    "- listed by the American College of Medical Genetics, `.germline.acmg.txt`\n",
    "- associated with somatic cancers, `.germline.cancer_related.txt`\n",
    "- and hereditary cancer risk, `.germline.hereditary_cancer.txt` \n",
    "\n",
    "Common variants and variants labeled as benign or likely benign by ClinVar are _not_ included in these outputs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b272dd19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>alteration_type</th>\n",
       "      <th>alteration</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start_position</th>\n",
       "      <th>end_position</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>observed_allele1</th>\n",
       "      <th>observed_allele2</th>\n",
       "      <th>...</th>\n",
       "      <th>exac_afr_an</th>\n",
       "      <th>exac_amr_an</th>\n",
       "      <th>exac_eas_an</th>\n",
       "      <th>exac_fin_an</th>\n",
       "      <th>exac_nfe_an</th>\n",
       "      <th>exac_sas_an</th>\n",
       "      <th>exac_oth_an</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>tumor_sample_barcode</th>\n",
       "      <th>normal_sample_barcode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Germline Variant</td>\n",
       "      <td>MSH6</td>\n",
       "      <td>Missense</td>\n",
       "      <td>p.G39E</td>\n",
       "      <td>2</td>\n",
       "      <td>48010488</td>\n",
       "      <td>48010488</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>6388.0</td>\n",
       "      <td>9014.0</td>\n",
       "      <td>6792.0</td>\n",
       "      <td>5078.0</td>\n",
       "      <td>48404.0</td>\n",
       "      <td>14790.0</td>\n",
       "      <td>664.0</td>\n",
       "      <td>example</td>\n",
       "      <td>example_tumor_profile</td>\n",
       "      <td>example_normal_profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Germline Variant</td>\n",
       "      <td>TP53</td>\n",
       "      <td>Missense</td>\n",
       "      <td>p.P72R</td>\n",
       "      <td>17</td>\n",
       "      <td>7579472</td>\n",
       "      <td>7579472</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>10030.0</td>\n",
       "      <td>11562.0</td>\n",
       "      <td>8620.0</td>\n",
       "      <td>6610.0</td>\n",
       "      <td>66692.0</td>\n",
       "      <td>16506.0</td>\n",
       "      <td>904.0</td>\n",
       "      <td>example</td>\n",
       "      <td>example_tumor_profile</td>\n",
       "      <td>example_normal_profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Germline Variant</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>Missense</td>\n",
       "      <td>p.N372H</td>\n",
       "      <td>13</td>\n",
       "      <td>32906729</td>\n",
       "      <td>32906729</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>9854.0</td>\n",
       "      <td>11568.0</td>\n",
       "      <td>8646.0</td>\n",
       "      <td>6614.0</td>\n",
       "      <td>66692.0</td>\n",
       "      <td>16490.0</td>\n",
       "      <td>904.0</td>\n",
       "      <td>example</td>\n",
       "      <td>example_tumor_profile</td>\n",
       "      <td>example_normal_profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Germline Variant</td>\n",
       "      <td>BRCA2</td>\n",
       "      <td>Frameshift</td>\n",
       "      <td>p.S1982fs</td>\n",
       "      <td>13</td>\n",
       "      <td>32914438</td>\n",
       "      <td>32914438</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>example</td>\n",
       "      <td>example_tumor_profile</td>\n",
       "      <td>example_normal_profile</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       feature_type feature alteration_type alteration  chromosome   \n",
       "0  Germline Variant    MSH6        Missense     p.G39E           2  \\\n",
       "1  Germline Variant    TP53        Missense     p.P72R          17   \n",
       "2  Germline Variant   BRCA2        Missense    p.N372H          13   \n",
       "3  Germline Variant   BRCA2      Frameshift  p.S1982fs          13   \n",
       "\n",
       "   start_position  end_position reference_allele observed_allele1   \n",
       "0        48010488      48010488                G                G  \\\n",
       "1         7579472       7579472                G                G   \n",
       "2        32906729      32906729                A                A   \n",
       "3        32914438      32914438                T                T   \n",
       "\n",
       "  observed_allele2  ...  exac_afr_an  exac_amr_an exac_eas_an  exac_fin_an   \n",
       "0                A  ...       6388.0       9014.0      6792.0       5078.0  \\\n",
       "1                C  ...      10030.0      11562.0      8620.0       6610.0   \n",
       "2                C  ...       9854.0      11568.0      8646.0       6614.0   \n",
       "3                -  ...          NaN          NaN         NaN          NaN   \n",
       "\n",
       "   exac_nfe_an  exac_sas_an  exac_oth_an  patient_id   tumor_sample_barcode   \n",
       "0      48404.0      14790.0        664.0     example  example_tumor_profile  \\\n",
       "1      66692.0      16506.0        904.0     example  example_tumor_profile   \n",
       "2      66692.0      16490.0        904.0     example  example_tumor_profile   \n",
       "3          NaN          NaN          NaN     example  example_tumor_profile   \n",
       "\n",
       "    normal_sample_barcode  \n",
       "0  example_normal_profile  \n",
       "1  example_normal_profile  \n",
       "2  example_normal_profile  \n",
       "3  example_normal_profile  \n",
       "\n",
       "[4 rows x 34 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = 'example-outputs/example.germline.acmg.txt'\n",
    "df = pd.read_csv(file, sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da7c35c7",
   "metadata": {},
   "source": [
    "Observed somatic nucleotide variants, germline variants, copy number alterations, and fusions will be combined by gene in the integrated summary output, `.integrated.summary.txt`, to facilitate integrative analyses. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "392a41e7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>somatic_mutations</th>\n",
       "      <th>copy_number_alterations</th>\n",
       "      <th>fusions</th>\n",
       "      <th>germline_mutations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ALK</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense p.I1461V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>BCR</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense p.N796S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BLM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Amplification</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>Missense p.V600E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Nonsense p.R509*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Missense p.N372H, Frameshift p.S1982fs</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  feature somatic_mutations copy_number_alterations fusions   \n",
       "0     ALK               NaN                     NaN     NaN  \\\n",
       "1     BCR               NaN                     NaN     NaN   \n",
       "2     BLM               NaN           Amplification     NaN   \n",
       "3    BRAF  Missense p.V600E                     NaN     NaN   \n",
       "4   BRCA2               NaN                     NaN     NaN   \n",
       "\n",
       "                       germline_mutations  \n",
       "0                       Missense p.I1461V  \n",
       "1                        Missense p.N796S  \n",
       "2                                     NaN  \n",
       "3                        Nonsense p.R509*  \n",
       "4  Missense p.N372H, Frameshift p.S1982fs  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = 'example-outputs/example.integrated.summary.txt'\n",
    "df = pd.read_csv(file, sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4142ebc8",
   "metadata": {},
   "source": [
    "Variants associated with microsatellite stability will also be extracted and summarized, `msi_variants.txt`. As of the current algorithm release, the following genes are aggregated,\n",
    "- _DOCK3_\n",
    "- _ESRP1_\n",
    "- _MLH1_\n",
    "- _MSH2_\n",
    "- _MSH6_\n",
    "- _PMS2_\n",
    "- _POLE_\n",
    "- _POLE2_\n",
    "- _PRDM2_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "e3fef139",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score_bin</th>\n",
       "      <th>sensitive_score_bin</th>\n",
       "      <th>resistance_score_bin</th>\n",
       "      <th>prognostic_score_bin</th>\n",
       "      <th>feature_type</th>\n",
       "      <th>feature</th>\n",
       "      <th>alteration_type</th>\n",
       "      <th>alteration</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start_position</th>\n",
       "      <th>...</th>\n",
       "      <th>patient_id</th>\n",
       "      <th>tumor_sample_barcode</th>\n",
       "      <th>normal_sample_barcode</th>\n",
       "      <th>almanac_bin</th>\n",
       "      <th>cancerhotspots_bin</th>\n",
       "      <th>cancerhotspots3D_bin</th>\n",
       "      <th>cgc_bin</th>\n",
       "      <th>gsea_pathways_bin</th>\n",
       "      <th>gsea_modules_bin</th>\n",
       "      <th>cosmic_bin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Biologically Relevant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Copy Number</td>\n",
       "      <td>JAK1</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>65047989</td>\n",
       "      <td>...</td>\n",
       "      <td>example</td>\n",
       "      <td>example_tumor_profile</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cancer Gene Census</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Germline Variant</td>\n",
       "      <td>PRDM2</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>p.E282del</td>\n",
       "      <td>1</td>\n",
       "      <td>14105122</td>\n",
       "      <td>...</td>\n",
       "      <td>example</td>\n",
       "      <td>example_tumor_profile</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               score_bin  sensitive_score_bin  resistance_score_bin   \n",
       "0  Biologically Relevant                  NaN                   NaN  \\\n",
       "1     Cancer Gene Census                  NaN                   NaN   \n",
       "\n",
       "   prognostic_score_bin      feature_type feature alteration_type alteration   \n",
       "0                   NaN       Copy Number    JAK1        Deletion        NaN  \\\n",
       "1                   NaN  Germline Variant   PRDM2        Deletion  p.E282del   \n",
       "\n",
       "   chromosome  start_position  ...  patient_id   tumor_sample_barcode   \n",
       "0           1        65047989  ...     example  example_tumor_profile  \\\n",
       "1           1        14105122  ...     example  example_tumor_profile   \n",
       "\n",
       "    normal_sample_barcode almanac_bin  cancerhotspots_bin   \n",
       "0  example_normal_profile           1                   1  \\\n",
       "1  example_normal_profile           0                   0   \n",
       "\n",
       "   cancerhotspots3D_bin  cgc_bin  gsea_pathways_bin  gsea_modules_bin   \n",
       "0                   0.0        1                1.0               1.0  \\\n",
       "1                   NaN        1                NaN               NaN   \n",
       "\n",
       "   cosmic_bin  \n",
       "0         1.0  \n",
       "1         NaN  \n",
       "\n",
       "[2 rows x 38 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = 'example-outputs/example.msi_variants.txt'\n",
    "df = pd.read_csv(file, sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3723457f",
   "metadata": {},
   "source": [
    "Tumor mutational burden will be calculated if somatic variants and the bases considered are provided, `.mutational_burden.txt`. Here, the nonsynyonmous mutational burden or coding mutational burden is calculated. Additionally, the Molecular Oncology Almanac will compare the observed mutational burden to TCGA.\n",
    "\n",
    "Tumor mutational burden will be evaluated as high if two conditions are met,\n",
    "- Mutations per megabase (Mb) > 10 \n",
    "- At least a mutational burden of 80th percentile of TCGA tumor type, if matched to a TCGA tumor type, else TCGA generally."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4ca533df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "patient_id                                          example\n",
       "reported_tumor_type                                    SKCM\n",
       "oncotree_term                            Cutaneous Melanoma\n",
       "oncotree_code                                          SKCM\n",
       "bases_covered                                    29908096.0\n",
       "n_nonsyn_mutations                                      204\n",
       "coding_mutational_burden_per_megabase                 6.821\n",
       "percentile_tcga                                       83.75\n",
       "percentile_tcga_tissuetype                            25.62\n",
       "high_burden?                                          False\n",
       "Name: 0, dtype: object"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = 'example-outputs/example.mutational_burden.txt'\n",
    "df = pd.read_csv(file, sep='\\t')\n",
    "df.loc[0, :]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a753763",
   "metadata": {},
   "source": [
    "Trinucleotide mutational signatures will be calculated using a subprocess if somatic variants are provided and the feature is enabled. The following relevant outputs are produced,\n",
    "- `.sigs.context.txt`, variant counts for each 96 trinucleotide bins\n",
    "- `.sigs.cosmic.txt`, weights for each of the COSMIC (v2) mutational signatures\n",
    "- `.sigs.tricontext.counts.png`, visualization of trinucleotide context counts of observed somatic variants\n",
    "- `.sigs.tricontext.normalized.png`, visualization of normalized counts of somatic variants across trinucleotide contexts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "f49ce4ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A[C&gt;A]A</th>\n",
       "      <th>A[C&gt;A]C</th>\n",
       "      <th>A[C&gt;A]G</th>\n",
       "      <th>A[C&gt;A]T</th>\n",
       "      <th>C[C&gt;A]A</th>\n",
       "      <th>C[C&gt;A]C</th>\n",
       "      <th>C[C&gt;A]G</th>\n",
       "      <th>C[C&gt;A]T</th>\n",
       "      <th>G[C&gt;A]A</th>\n",
       "      <th>G[C&gt;A]C</th>\n",
       "      <th>...</th>\n",
       "      <th>C[T&gt;G]G</th>\n",
       "      <th>C[T&gt;G]T</th>\n",
       "      <th>G[T&gt;G]A</th>\n",
       "      <th>G[T&gt;G]C</th>\n",
       "      <th>G[T&gt;G]G</th>\n",
       "      <th>G[T&gt;G]T</th>\n",
       "      <th>T[T&gt;G]A</th>\n",
       "      <th>T[T&gt;G]C</th>\n",
       "      <th>T[T&gt;G]G</th>\n",
       "      <th>T[T&gt;G]T</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 96 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   A[C>A]A  A[C>A]C  A[C>A]G  A[C>A]T  C[C>A]A  C[C>A]C  C[C>A]G  C[C>A]T   \n",
       "0        0        2        1        0        3        1        1        1  \\\n",
       "\n",
       "   G[C>A]A  G[C>A]C  ...  C[T>G]G  C[T>G]T  G[T>G]A  G[T>G]C  G[T>G]G   \n",
       "0        0        0  ...        0        3        0        0        0  \\\n",
       "\n",
       "   G[T>G]T  T[T>G]A  T[T>G]C  T[T>G]G  T[T>G]T  \n",
       "0        0        0        1        0        0  \n",
       "\n",
       "[1 rows x 96 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = 'example-outputs/example.sigs.context.txt'\n",
    "df = pd.read_csv(file, sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dd44b4e",
   "metadata": {},
   "source": [
    "<img src=\"example-outputs/example.sigs.tricontext.counts.png\" alt=\"Variant counts by trinucleotide context\" width=\"750\"/>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d304a7db",
   "metadata": {},
   "source": [
    "<a href=\"#toc\">Return to Table of Contents</a>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb903b79",
   "metadata": {},
   "source": [
    "<a id='conclusion'></a>\n",
    "## In conclusion"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10c8b3c2",
   "metadata": {},
   "source": [
    "**Thank you very much for inviting us to give this workshop today, please reach out with any additional questions or comments!**\n",
    "\n",
    "- Email: moalmanac@ds.dfci.harvard.edu\n",
    "- Twitter: [@moalmanac](https://twitter.com/moalmanac), [@vanallenlab](https://twitter.com/vanallenlab)\n",
    "\n",
    "Slides and code from this workshop are available on GitHub at [https://github.com/vanallenlab/2023-cgc-moalmanac](https://github.com/vanallenlab/2023-cgc-moalmanac)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff2a6dbf",
   "metadata": {},
   "source": [
    "<img src=\"img/conclusion-slide.png\" alt=\"Thank you!\" width=\"750\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14336891",
   "metadata": {},
   "source": [
    "Molecular Oncology Almanac (MOAlmanac) is a clinical interpretation algorithm paired with an underlying knowledge base for precision oncology. The primary objective of MOAlmanac is to identify and associate molecular alterations with therapeutic sensitivity and resistance as well as disease prognosis. This is done for “first-order” genomic alterations -- individual events such as somatic variants, copy number alterations, fusions, and germline -- as well as “second-order” events -- those that are not associated with one single mutation, and may be descriptive of global processes in the tumor such as tumor mutational burden, microsatellite instability, mutational signatures, and whole-genome doubling. In addition to clinical insights, MOAlmanac will annotate and evaluate first-order events based on their presence in numerous other well established datasources as well as highlight connections between them. This method currently geared towards hg19/b37 reference files and whole-exome or targeted sequencing data.\n",
    "\n",
    "There are several other resources within [the Molecular Oncology Almanac ecosystem](https://github.com/topics/molecular-oncology-almanac): \n",
    "- Molecular Oncology Almanac, the clinical interpretation algorithm for precision cancer medicine [[GitHub](https://github.com/vanallenlab/moalmanac)].\n",
    "- [Molecular Oncology Almanac Browser](https://moalmanac.org), a website to browse our underlying knowledge base [[GitHub](https://github.com/vanallenlab/moalmanac-browser)].\n",
    "- [Molecular Oncology Almanac Connector](https://chrome.google.com/webstore/detail/molecular-oncology-almana/jliaipolchffpaccagodphgjpfdpcbcm?hl=en), a Google Chrome extension to quickly suggest literature for cataloging [[GitHub](https://github.com/vanallenlab/moalmanac-extension)].\n",
    "- Molecular Oncology Almanac Database, the content and release notes of our underlying knowledge base [[GitHub](https://github.com/vanallenlab/moalmanac-db)].\n",
    "- [Molecular Oncology Almanac Portal](https://portal.moalmanac.org), a website to launch this method on the Broad Institute's Google Cloud platform called Terra [[GitHub](https://github.com/vanallenlab/moalmanac-portal)].\n",
    "\n",
    "This method is also available on [Docker](https://hub.docker.com/repository/docker/vanallenlab/moalmanac) and [Terra](https://portal.firecloud.org/#methods/vanallenlab/moalmanac/). We have also released [code on GitHub to help facilitate analyses of multiple samples that have been interpreted using the Molecular Oncology Almanac](https://github.com/vanallenlab/moalmanac-cohort). \n",
    "\n",
    "If you use this method, please cite our publication:\n",
    "> [Reardon, B., Moore, N.D., Moore, N.S., *et al*. Integrating molecular profiles into clinical frameworks through the Molecular Oncology Almanac to prospectively guide precision oncology. *Nat Cancer* (2021). https://doi.org/10.1038/s43018-021-00243-3](https://www.nature.com/articles/s43018-021-00243-3)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "2023-cgc-workshop-moalmanac",
   "language": "python",
   "name": "2023-cgc-workshop-moalmanac"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
