{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Manufacture germline input\n",
    "For testing and illustrative purposes we do not want to use any actual germline data. Instead, we just make a list of some known common variants and some which we anticipate to be highlighted by MOAlmanac. While most germline samples should have on the order of tens of thousands of variants, we just manually manufacture a file of 10. \n",
    "\n",
    "Some items for simplicity\n",
    "- All alt and ref counts will be 50 and 50\n",
    "- We only consider SNPs, so start position will always equal the end position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "gene = 'Hugo_Symbol'\n",
    "ncbi = 'NCBI_Build'\n",
    "chrom = 'Chromosome'\n",
    "start = 'Start_position'\n",
    "end = 'End_position'\n",
    "v_class = 'Variant_Classification'\n",
    "ref = 'Reference_Allele'\n",
    "obs1 = 'Tumor_Seq_Allele1'\n",
    "obs2 = 'Tumor_Seq_Allele2'\n",
    "tumor = 'Tumor_Sample_Barcode'\n",
    "normal = 'Matched_Norm_Sample_Barcode'\n",
    "transcript = 'Annotation_Transcript'\n",
    "protein = 'Protein_Change'\n",
    "alt_count = 't_alt_count'\n",
    "ref_count = 't_ref_count'\n",
    "\n",
    "columns = [gene, ncbi, chrom, start, end, v_class, ref, obs1, obs2, \n",
    "           tumor, normal, transcript, protein, alt_count, ref_count]\n",
    "\n",
    "missense = 'Missense_Mutation'\n",
    "frameshift = 'Frame_Shift_Del'\n",
    "deletion = 'In_Frame_Del'\n",
    "nonsense = 'Nonsense_Mutation'\n",
    "\n",
    "build_value = 37\n",
    "tumor_value = '__UNKNOWN__'\n",
    "normal_value = 'example_normal_profile'\n",
    "transcript_value = ''\n",
    "alt_count_value = 50\n",
    "ref_count_value = 50\n",
    "\n",
    "outname = 'example_patient.capture.germline.maf'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hugo_Symbol</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>Tumor_Seq_Allele1</th>\n",
       "      <th>Tumor_Seq_Allele2</th>\n",
       "      <th>Tumor_Sample_Barcode</th>\n",
       "      <th>Matched_Norm_Sample_Barcode</th>\n",
       "      <th>Annotation_Transcript</th>\n",
       "      <th>Protein_Change</th>\n",
       "      <th>t_alt_count</th>\n",
       "      <th>t_ref_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PRDM2</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>14105122</td>\n",
       "      <td>14105122</td>\n",
       "      <td>In_Frame_Del</td>\n",
       "      <td>GAA</td>\n",
       "      <td>GAA</td>\n",
       "      <td>-</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td></td>\n",
       "      <td>p.E282del</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ALK</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>29416572</td>\n",
       "      <td>29416572</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td></td>\n",
       "      <td>p.I1461V</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BIRC6</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>32667182</td>\n",
       "      <td>32667182</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td></td>\n",
       "      <td>p.V1332L</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSH6</td>\n",
       "      <td>37</td>\n",
       "      <td>2</td>\n",
       "      <td>48010488</td>\n",
       "      <td>48010488</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td></td>\n",
       "      <td>p.G39E</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FGFR4</td>\n",
       "      <td>37</td>\n",
       "      <td>5</td>\n",
       "      <td>176520243</td>\n",
       "      <td>176520243</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td></td>\n",
       "      <td>p.G388R</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>BRAF</td>\n",
       "      <td>37</td>\n",
       "      <td>7</td>\n",
       "      <td>140476881</td>\n",
       "      <td>140476881</td>\n",
       "      <td>Nonsense_Mutation</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td></td>\n",
       "      <td>p.R509*</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TP53</td>\n",
       "      <td>37</td>\n",
       "      <td>17</td>\n",
       "      <td>7579472</td>\n",
       "      <td>7579472</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>G</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td></td>\n",
       "      <td>p.P72R</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>37</td>\n",
       "      <td>13</td>\n",
       "      <td>32906729</td>\n",
       "      <td>32906729</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td></td>\n",
       "      <td>p.N372H</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>BRCA2</td>\n",
       "      <td>37</td>\n",
       "      <td>13</td>\n",
       "      <td>32914438</td>\n",
       "      <td>32914438</td>\n",
       "      <td>Frame_Shift_Del</td>\n",
       "      <td>T</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td></td>\n",
       "      <td>p.S1982fs</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>BCR</td>\n",
       "      <td>37</td>\n",
       "      <td>22</td>\n",
       "      <td>23627369</td>\n",
       "      <td>23627369</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>A</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>__UNKNOWN__</td>\n",
       "      <td>example_normal_profile</td>\n",
       "      <td></td>\n",
       "      <td>p.N796S</td>\n",
       "      <td>50</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hugo_Symbol  NCBI_Build  Chromosome  Start_position  End_position  \\\n",
       "0       PRDM2          37           1        14105122      14105122   \n",
       "1         ALK          37           2        29416572      29416572   \n",
       "2       BIRC6          37           2        32667182      32667182   \n",
       "3        MSH6          37           2        48010488      48010488   \n",
       "4       FGFR4          37           5       176520243     176520243   \n",
       "5        BRAF          37           7       140476881     140476881   \n",
       "6        TP53          37          17         7579472       7579472   \n",
       "7       BRCA2          37          13        32906729      32906729   \n",
       "8       BRCA2          37          13        32914438      32914438   \n",
       "9         BCR          37          22        23627369      23627369   \n",
       "\n",
       "  Variant_Classification Reference_Allele Tumor_Seq_Allele1 Tumor_Seq_Allele2  \\\n",
       "0           In_Frame_Del              GAA               GAA                 -   \n",
       "1      Missense_Mutation                T                 T                 C   \n",
       "2      Missense_Mutation                G                 G                 C   \n",
       "3      Missense_Mutation                G                 G                 A   \n",
       "4      Missense_Mutation                G                 G                 A   \n",
       "5      Nonsense_Mutation                G                 G                 A   \n",
       "6      Missense_Mutation                G                 G                 C   \n",
       "7      Missense_Mutation                A                 A                 C   \n",
       "8        Frame_Shift_Del                T                 T                 -   \n",
       "9      Missense_Mutation                A                 A                 G   \n",
       "\n",
       "  Tumor_Sample_Barcode Matched_Norm_Sample_Barcode Annotation_Transcript  \\\n",
       "0          __UNKNOWN__      example_normal_profile                         \n",
       "1          __UNKNOWN__      example_normal_profile                         \n",
       "2          __UNKNOWN__      example_normal_profile                         \n",
       "3          __UNKNOWN__      example_normal_profile                         \n",
       "4          __UNKNOWN__      example_normal_profile                         \n",
       "5          __UNKNOWN__      example_normal_profile                         \n",
       "6          __UNKNOWN__      example_normal_profile                         \n",
       "7          __UNKNOWN__      example_normal_profile                         \n",
       "8          __UNKNOWN__      example_normal_profile                         \n",
       "9          __UNKNOWN__      example_normal_profile                         \n",
       "\n",
       "  Protein_Change  t_alt_count  t_ref_count  \n",
       "0      p.E282del           50           50  \n",
       "1       p.I1461V           50           50  \n",
       "2       p.V1332L           50           50  \n",
       "3         p.G39E           50           50  \n",
       "4        p.G388R           50           50  \n",
       "5        p.R509*           50           50  \n",
       "6         p.P72R           50           50  \n",
       "7        p.N372H           50           50  \n",
       "8      p.S1982fs           50           50  \n",
       "9        p.N796S           50           50  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variants = [\n",
    "    {gene:'PRDM2', chrom:1, start:14105122, ref:'GAA', obs1:'GAA', obs2:'-', v_class:deletion, protein:'p.E282del'},\n",
    "    {gene:'ALK', chrom:2, start:29416572, ref:'T', obs1:'T', obs2:'C', v_class:missense, protein:'p.I1461V'},\n",
    "    {gene:'BIRC6', chrom:2, start:32667182, ref:'G', obs1:'G', obs2:'C', v_class:missense, protein:'p.V1332L'},\n",
    "    {gene:'MSH6', chrom:2, start:48010488, ref:'G', obs1:'G', obs2:'A', v_class:missense, protein:'p.G39E'},\n",
    "    {gene:'FGFR4', chrom:5, start:176520243, ref:'G', obs1:'G', obs2:'A', v_class:missense, protein:'p.G388R'},\n",
    "    {gene:'BRAF', chrom:7, start:140476881, ref:'G', obs1:'G', obs2:'A', v_class:nonsense, protein:'p.R509*'},\n",
    "    {gene:'TP53', chrom:17, start:7579472, ref:'G', obs1:'G', obs2:'C', v_class:missense, protein:'p.P72R'},\n",
    "    {gene:'BRCA2', chrom:13, start:32906729, ref:'A', obs1:'A', obs2:'C', v_class:missense, protein:'p.N372H'},\n",
    "    {gene:'BRCA2', chrom:13, start:32914438, ref:'T', obs1:'T', obs2:'-', v_class:frameshift, protein:'p.S1982fs'},\n",
    "    {gene:'BCR', chrom:22, start:23627369, ref:'A', obs1:'A', obs2:'G', v_class:missense, protein:'p.N796S'},\n",
    "]\n",
    "\n",
    "df = pd.DataFrame.from_records(variants)\n",
    "\n",
    "for column in columns:\n",
    "    if column not in df.columns:\n",
    "        df[column] = ''\n",
    "df = df.loc[:, columns]        \n",
    "\n",
    "pairs = [(ncbi, build_value), (tumor, tumor_value), (normal, normal_value), (transcript, transcript_value), (alt_count, alt_count_value), (ref_count, ref_count_value)]\n",
    "for column, value in pairs:\n",
    "    df[column] = value\n",
    "\n",
    "df[end] = df[start]\n",
    "\n",
    "df.to_csv(outname, sep='\\t', index=False)\n",
    "\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "moalmanac",
   "language": "python",
   "name": "moalmanac"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
